design/dedicated-osd-pod.md

```diff
@@ -6,53 +6,97 @@ A one-OSD-per-Pod placement should be implemented to improve reliability and res
+
 ## Proposal   
 
-Based on offline discussion with [Travis](https://github.com/travisn), we propose the following OSD startup change.
+We propose the following change to address the limitations.
+
 
 ### Create new OSDs
-| Sequence |Rook Operator  | Rook OSD Agent  | Ceph OSD  | Note  |
+| Sequence |Rook Operator | Device Discovery  | Device Provisioner   | Ceph OSD Deployment | 
 |---|---|---|---|---|
-| 1  | Create a Pod for OSD Agents for each storage node  |   |   | Consider a CronJob for peridically invocation to detect device change  |
-| 2  |   |  Discover and prepare OSDs  |   | |
-| 3  |   |  Persist OSD ID, datapath, and node info in a per node Configmap | | |
-| 4  | Watch Configmap, parse Configmap, extract OSD info, construct OSD Pod command and arg | | |
-| 5  | Create one deployment per OSD | | |
-| 6  | | | Start OSD Daemon | Running `ceph-osd` in a dedicated Pod eliminates the rules and caps assigned to Rook OSD Agent |
+| 1  | Read devices from cluster CRD and create an OSD Discovery deployment on each storage node  |   |   |
+| 2  | Watch raw device Configmap  | Discover unformatted devices and update raw device Configmap with the device paths and device information (size, type, etc)| | |
+| 3  | Detect raw device Configmap change, parse Configmap, extract device paths with matching filters defined in CRD, and create an Device Provisioner deployment for each device  || | |
+| 4  | Watch device provisioning Configmap | | Prepare OSDs, Persist OSD ID, datapath, and node info in a per node device provisioning Configmap | |
+| 5  | Detect device provisioning Configmap change, parse Configmap, extract OSD info, construct OSD Pod command and arg | | | |
+| 6  | Create one deployment per OSD | | |Start `ceph-osd` Daemon one Pod per device |
+
+
+This change addresses the above limitations in the follow ways:
+- High reliability. Each `ceph-osd` daemon runs its own Pod, their restart and upgrade are by Kubernetes controllers. Upgrading Device Provisioner Pod no longer restarts `ceph-osd` daemons.
+- More efficient resource requests. Once Device Discovery detects all devices, Rook Operator is informed of the topology and assigns appropriate resources to each Ceph OSD deployment.
+- Reusable. Device discovery can be used for other storage backends.
+
+
+### Detailed Device Discovery Process
 
-### OSD Addition/Deletion
+Each `Device Discovery` Pod walks through device trees to unformatted block devices and stores the device information in `Raw Device Configmap`. 
 
-Rook OSD agent updates Configmap to reflect current OSD devices on the node.
-Rook Operator creates/deletes OSD deployments based on Configmap change. 
+A sample of `Raw Device Configmap` is as the following:
 
-### OSD Pod Naming
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  namespace: rook-system
+  name: raw-devices
+data:
+    - node: "node1"
```

Comment:
Many device discovery pods will be running in the cluster and will all need to update this configmap, which would cause contention. It seems we would need to have a separate configmap for each node to avoid the contention.